// @aws-sdk/client-apigatewaymanagementapi-node/commands/PostToConnectionCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var __aws_sdk_middleware_stack = require("@aws-sdk/middleware-stack");
var PostToConnection_1 = require("../model/operations/PostToConnection");
var PostToConnectionCommand = /** @class */ (function () {
    function PostToConnectionCommand(input) {
        this.input = input;
        this.model = PostToConnection_1.PostToConnection;
        this.middlewareStack = new __aws_sdk_middleware_stack.MiddlewareStack();
    }
    PostToConnectionCommand.prototype.resolveMiddleware = function (clientStack, configuration) {
        var handler = configuration.handler;
        var stack = clientStack.concat(this.middlewareStack);
        var handlerExecutionContext = {
            logger: {},
            model: this.model
        };
        return stack.resolve(handler(handlerExecutionContext), handlerExecutionContext);
    };
    return PostToConnectionCommand;
}());
exports.PostToConnectionCommand = PostToConnectionCommand;
